load-testing/tests/relays_stress_helpers_test.go

```diff
@@ -0,0 +1,1143 @@
+		if waitingForFirstSession && blockHeight != sessionInfo.sessionStartBlockHeight {
+			countDownToTestStart := sessionInfo.sessionEndBlockHeight - blockHeight + 1
+			infoLogger.Msgf(
+				"waiting for next session to start: in %d blocks",
+				countDownToTestStart,
+			)
+
+			// The test is not to be started yet, skip the notification to the downstream
+			// observables until the first block of the next session is reached.
+			return nil, true
+		}
+
+		// If the test has not started, set the start block height to the current block height.
+		// As soon as the test start, s.startBlockHeight will no longer be updated.
+		// It is updated only once at the start of the test.
+		if waitingForFirstSession {
+			s.startBlockHeight = blockHeight
+		}
+
+		// Mark the test as started.
+		waitingForFirstSession = false
+
+		// Log the test progress.
+		infoLogger.Msgf(
+			"test progress blocks: %d/%d",
+			blockHeight-s.startBlockHeight, s.testDurationBlocks,
+		)
+
+		// If the test duration is reached, cancel the context to stop the test.
+		if blockHeight >= s.startBlockHeight+s.testDurationBlocks {
+			logger.Info().Msg("Test done, cancelling scenario context")
+			s.cancelCtx()
+
+			return nil, true
+		}
+
+		// If the current block is the start of any new session, activate the prepared
+		// actors to be used in the current session.
+		s.activatePreparedActors(sessionInfo)
+
+		now := time.Now()
+
+		// Inform the relay sending observable of the active applications that
+		// will be sending relays and the gateways that will be receiving them.
+		batchInfoPublishCh <- &batchInfoNotif{
+			sessionInfoNotif: *sessionInfo,
+			prevBatchTime:    prevBatchTime,
+			nextBatchTime:    now,
+			appAccounts:      s.activeApplications,
+			gateways:         s.activeGateways,
+		}
+
+		// Update prevBatchTime after this iteration completes.
+		prevBatchTime = now
+
+		// Forward the session info notification to the downstream observables.
+		return sessionInfo, false
+	}
+}
+
+// TODO_TECHDEBT: godoc comment.
+func (s *relaysSuite) validateActorPlans(plans *actorPlans) {
+	plans.validateAppSupplierPermutations(s)
+	plans.validateIncrementRates(s)
+	plans.validateMaxAmounts(s)
+
+	require.Truef(s,
+		len(s.gatewayUrls) >= int(plans.gateways.maxAmount),
+		"provisioned gateways must be greater or equal than the max gateways to be staked",
+	)
+	require.Truef(s,
+		len(s.suppliersUrls) >= int(plans.suppliers.maxAmount),
+		"provisioned suppliers must be greater or equal than the max suppliers to be staked",
+	)
+}
+
+// TODO_IN_THIS_COMMIT: godoc comment.
+func (plans *actorPlans) maxDurationBlocks() int64 {
+	return math.Max(
+		plans.gateways.durationBlocks(),
+		plans.apps.durationBlocks(),
+		plans.suppliers.durationBlocks(),
+	)
+}
+
+func (plans *actorPlans) validateAppSupplierPermutations(t gocuke.TestingT) {
+	// Ensure that the number of suppliers never goes above the number of applications.
```

Comment:
How do we ensure that we don't end up having `app1:sup1`, `app2:sup1`, `app3:sup1`, with `sup2`,`sup3`, etc unused?

I feel like this is more a function of the NumSuppliersPerSession rather than the app:supplier ratio, no?

Reply:
![image](https://github.com/pokt-network/poktroll/assets/600733/a4c4eed6-d8db-4e2f-bb3c-363db9262e90)